stellargraph/layer/appnp.py

```diff
@@ -0,0 +1,244 @@
+        the paper (alpha in the paper)
+        approx_iter: number of iterations to approximate PPNP as described in the paper (K in the paper)
+    """
+
+    def __init__(
+            self,
+            layer_sizes,
+            activations,
+            generator,
+            bias=True,
+            dropout=0.0,
+            transport_probability=0.1,
+            kernel_regularizer=None,
+            approx_iter=10
+    ):
+
+        if not isinstance(generator, FullBatchNodeGenerator):
+            raise TypeError("Generator should be a instance of FullBatchNodeGenerator")
+
+        assert len(layer_sizes) == len(activations)
+
+        self.layer_sizes = layer_sizes
+        self.transport_probability = transport_probability
+        self.activations = activations
+        self.bias = bias
+        self.dropout = dropout
+        self.kernel_regularizer = kernel_regularizer
+        self.generator = generator
+        self.support = 1
+        self.method = generator.method
+
+        # Check if the generator is producing a sparse matrix
+        self.use_sparse = generator.use_sparse
+        if self.method == "none":
+            self.graph_norm_layer = GraphPreProcessingLayer(
+                num_of_nodes=self.generator.Aadj.shape[0]
+            )
+
+        # Initialize a stack of fully connected layers
+        n_layers = len(self.layer_sizes)
+        self._layers = []
+        for ii in range(n_layers):
+            l = self.layer_sizes[ii]
+            a = self.activations[ii]
+            self._layers.append(Dropout(self.dropout))
+            self._layers.append(
+                Dense(
+                    l,
+                    activation=a,
+                    use_bias=self.bias,
+                    kernel_regularizer=self.kernel_regularizer,
+                )
+            )
+
+        feature_dim = self.layer_sizes[-1]
+
+        for ii in range(approx_iter):
+            self._layers.append(Dropout(self.dropout))
+            #use graph convolutions to perform the AH matrix multiplication (A = adjacency matrix,
+            # H = previous layer feature matrix)
+            #set the weights W = I (identity), bias=0, activation=linear, and make the layer non-trainable
+            # so the convolution becomes: activation(AHW + bias) = (AHI + 0) = AH
+            #add propogation step with lambda function in the __call__ function
+            self._layers.append(
+                GraphConvolution(
+                    feature_dim,
+                    activation='linear',
+                    use_bias=False,
+                    kernel_initializer='Identity',
+                    final_layer=False,
+                    trainable=False
+                )
+            )
+        #apply softmax
+        self._layers.append(Softmax())
+
+        #gather the nodes from the output indices - inputs = [all node predictions, output node indices]
+        self._layers.append(
```

Comment:
If you specify in the final graph convolutional layer `final_layer=True` it will gather the outputs according to the output indices without requiring this final Lambda layer. Or am I missing something?

Reply:
I have to do one last addition of the feature layer after the final convolution - so the gathering operation has to be done after the final addition not the final convolution. I actually tried this earlier by excluding the final addition of the feature layer but this tanked the accuracy!

Reply:
I see, in that case a new matrix multiplication layer with the added teleport will make this neater.